File: 3020-find-the-maximum-number-of-elements-in-subset/solution.py
from collections import Counter
import math

class Solution:
    def maximumLength(self, nums: list[int]) -> int:
        counts = Counter(nums)
        max_len = 1
        
        # Handle base 1
        if 1 in counts:
            ones = counts[1]
            if ones % 2 == 0:
                ones -= 1
            max_len = max(max_len, ones)
            
        # Handle bases > 1
        for x in counts:
            if x == 1:
                continue
            
            curr = x
            current_len = 0
            
            # Count pairs for each power
            while counts.get(curr, 0) >= 2:
                current_len += 2
                curr *= curr
            
            # Check if we can add a middle element
            if counts.get(curr, 0) >= 1:
                current_len += 1
            else:
                # If we couldn't form the last pair, remove the overshoot
                current_len -= 1
            
            max_len = max(max_len, current_len)
            
        return max_len

File: 3020-find-the-maximum-number-of-elements-in-subset/solution.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <cmath>

class Solution {
public:
    int maximumLength(std::vector<int>& nums) {
        std::unordered_map<long long, int> counts;
        for (int num : nums) {
            counts[num]++;
        }

        int max_len = 1;

        // Handle base 1 separately
        if (counts.count(1)) {
            int ones = counts[1];
            if (ones % 2 == 0) ones--;
            max_len = std::max(max_len, ones);
        }

        // Handle bases > 1
        for (auto const& [x, count] : counts) {
            if (x == 1) continue;

            long long curr = x;
            int current_len = 0;
            
            while (counts.count(curr) && counts[curr] >= 2) {
                current_len += 2;
                curr *= curr;
            }
            
            if (counts.count(curr)) {
                current_len += 1;
            } else {
                // If we ran out, the last squared value was actually the middle
                current_len -= 1;
            }
            
            max_len = std::max(max_len, current_len);
        }

        return max_len;
    }
};